### PR TITLE
write asynchronously when handshaking

### PR DESCRIPTION
**Pull request by Stebalien**
Given that *both* sides now try to write the multistream handshake out before reading, we need to do it asynchronously. Otherwise, we can deadlock (depending on the protocol).

In practice, this shouldn't be an issue but I'm not sure if we can really *guarantee* that.

This fix *does* cost us an ephemeral goroutine but I'm not sure if we can get away from that.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ func SelectOneOf(protos []string, rwc io.ReadWriteCloser) (string, error) {
 }
 
 func handshake(rwc io.ReadWriteCloser) error {
-	if err := delimWrite(rwc, []byte(ProtocolID)); err != nil {
+	if err := delimWriteBuffered(rwc, []byte(ProtocolID)); err != nil {
```
Comment (Stebalien) [inline client.go]: Drive-by fix to reduce the number of packets we send.

**Comment by Stebalien** [comment]
Note: I hit this when testing with a pipe (in a different package).

**Comment by whyrusleeping** [comment]
I initially wrote things this way to eplicitly avoid having an extra goroutine like this. But i guess 'goroutines are cheap', and we're opening streams way less often than we used to (though you want to change things mr @Stebalien )

**Comment by marten-seemann** [comment]
Why could we deadlock when both peers are writing to the stream at the same time? Which stream muxers does this apply to?

**Comment by Stebalien** [comment]
@whyrusleeping we'll be using a different protocol (where the client will always speak first). Here:

1. In the past, the server spoke first. That doesn't work with protocols like QUIC that don't "open" the stream until data is sent.
2. Now, they both speak at the same time. We can't change the *server's* behavior as that'd break old clients.

@marten-seemann 

> Why could we deadlock when both peers are writing to the stream at the same time? Which stream muxers does this apply to?

The problem is that we aren't reading. We could also just assume that there's a small write buffer.

**Comment by Stebalien** [comment]
Note: the *lazy* negotiation streams (the fast path ones) already launch goroutines so I'm not so concerned here.

**Comment by marten-seemann** [comment]
I think it’s a valid assumption that we have initially have enough flow control credit to write a protocol ID, so I’d prefer to keep things simple here.

**Comment by Stebalien** [comment]
> I think it’s a valid assumption that we have initially have enough flow control credit to write a protocol ID...

This is more about buffering than control flow credit. For example, in a theoretical (but planned) "local" transport, each "stream" would just be a `net.Pipe`. We could have a buffering requirement but I'd rather not, especially because `net.Pipe` doesn't and go had rejected proposals to *add* a buffer multiple times.

File: client.go [hunk 2]
```diff
@@ -44,13 +44,19 @@ func SelectOneOf(protos []string, rwc io.ReadWriteCloser) (string, error) {
 }
 
 func handshake(rwc io.ReadWriteCloser) error {
-	if err := delimWrite(rwc, []byte(ProtocolID)); err != nil {
-		return err
-	}
+	errCh := make(chan error)
```
Comment (whyrusleeping) [inline client.go]: Thoughts on making this buffered? I know theres no early outs, but if `ReadNextToken` panics for some reason it could leave a hanging goroutine. 
Probably not a big deal though

Comment (Stebalien) [reply client.go]: We don't catch the panic, so we'll just crash the program anyways.

**Comment by Stebalien** [comment]
I'm going with this for now. It's the conservative approach and we can loosen it up if we find performance issues.

File: client.go (hunk 2 above)
Comment (Stebalien) [reply client.go]: Oh, *I* see. Hm. I'm always wary about recovering from panics, but you're probably right.

Comment (Stebalien) [reply client.go]: Fixed: https://github.com/multiformats/go-multistream/pull/30
